### PR TITLE
feat!: Add support for creating an associated dead-letter queue and queue policies

### LINKED ISSUE #44: Queue is created with SSE enabled on first run
## Description

On the first apply, queue (in our case FIFO queue) is created with Server Side Encryption enabled (using SQS managed key) even though the default should be disabled SSE.
Re-applying the plan detects drift and disables SSE. 

## ⚠️ Note

Before you submit an issue, please perform the following first:

1. Remove the local `.terraform` directory (! ONLY if state is stored remotely, which hopefully you are following that best practice!): `rm -rf .terraform/`
2. Re-initialize the project root to pull down modules: `terraform init`
3. Re-attempt your terraform plan or apply and check if the issue still persists

## Versions

- Module version [Required]: 3.4.2

- Terraform version: 1.3.4
- Provider version(s): 1.3.4

## Reproduction Code [Required]

```terraform
module "my_queue" {
  source  = "terraform-aws-modules/sqs/aws"
  version = "3.4.2"

  name = "my-queue.fifo"
  
  fifo_queue = true
  content_based_deduplication = true
}
```

## Expected behavior

SQS queue with disabled SSE is created.

## Actual behavior

SQS queue with SSE enabled (using SQS managed key) is created.

### PR DESCRIPTION
**Pull request by bryantbiggs**
## Description
- `var.redrive_policy`, and `var.redrive_policy` data types have changed from `string` to `any` which is a map of values. The conversion from a map to a jsonencoded string is now handled by the module
- `var.name_prefix` has been replaced with `var.use_name_prefix` which is a boolean that will use `var.name` as a prefix
- `var.policy` has been removed; users can create a policy via the queue policy or dead-letter queue policy which by default uses the associated queue ARN as the `resource` (avoids the chicken vs the egg scenario)
- When creating a FIFO queue, the `.fifo` postfix will now be automatically added to the queue name
- Added support for creating:
  - Queue policy
  - Dead letter queue
  - Dead letter queue policy
- Redrive and redrive allow policies have been converted to their separate resources to avoid lifecycle conflicts; now you can create both the source queue and dead-letter queue in the same `terraform apply` without conflict
- The queue data source previously used to extract the queue name has been replaced since this is natively supported in the current AWS provider queue resource

Please see the [upgrade guide](https://github.com/bryantbiggs/terraform-aws-sqs/blob/feat/sqs-dlq/UPGRADE-4.0.md) for full list of changes

## Motivation and Context
- It is quite common, and recommended, to create a dead-letter queue for messages that fail processing. This change makes that configuration possible and easier for users
- Using queue policies can create a chick vs the egg scenario where the policy needs the ARN of the queue but the queue requires the policy. Creating the policy within the queue module, alongside the resource, alleviates the situation while also allowing users to configure a policy that best suites their requirements
- The changes here have been tested with the changes proposed in https://github.com/terraform-aws-modules/terraform-aws-sns/pull/36 to ensure that the common pub-sub pattern that can be created by the two modules, with support for DLQ, works without conflict. Once this change is finalized and merged, I will update that PR since it uses this module in the example for pub-sub
- Resolves #44 
- Resolves #45

## Breaking Changes
- Yes; please see the [upgrade guide](https://github.com/bryantbiggs/terraform-aws-sqs/blob/feat/sqs-dlq/UPGRADE-4.0.md)

## How Has This Been Tested?
- [x] I have updated at least one of the `examples/*` to demonstrate and validate my change(s)
- [x] I have tested and validated these changes using one or more of the provided `examples/*` projects
- [x] I have executed `pre-commit run -a` on my pull request

### CODE REVIEW

File: UPGRADE-4.0.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
+# Upgrade from v3.x to v4.x
+
+If you have any questions regarding this upgrade process, please consult the [`examples`](https://github.com/terraform-aws-modules/terraform-aws-sns/tree/master/examples/complete) directory:
+
+If you find a bug, please open an issue with supporting configuration to reproduce.
+
+## List of backwards incompatible changes
+
+- `var.redrive_policy`, and `var.redrive_policy` data types have changed from `string` to `any` which is a map of values. The conversion from a map to a jsonencoded string is now handled by the module
```
Comment (bryantbiggs) [inline UPGRADE-4.0.md:9]:
```suggestion
- `var.redrive_policy`, and `var.redrive_allow_policy` data types have changed from `string` to `any` which is a map of values. The conversion from a map to a jsonencoded string is now handled by the module
```

File: main.tf [hunk 2]
```diff
@@ -1,32 +1,225 @@
...
+  content_based_deduplication       = var.content_based_deduplication
+  deduplication_scope               = var.deduplication_scope
+  delay_seconds                     = var.delay_seconds
+  fifo_queue                        = var.fifo_queue
+  fifo_throughput_limit             = var.fifo_throughput_limit
   kms_data_key_reuse_period_seconds = var.kms_data_key_reuse_period_seconds
+  kms_master_key_id                 = var.kms_master_key_id
+  max_message_size                  = var.max_message_size
+  message_retention_seconds         = var.message_retention_seconds
+  name                              = var.use_name_prefix ? null : (var.fifo_queue ? "${trimsuffix(var.name, ".fifo")}.fifo" : var.name)
```
Comment (antonbabenko) [inline main.tf:17]: 👍 `trimsuffix`

**Review by antonbabenko — APPROVED** [review]

**Comment by antonbabenko** [comment]
This PR is included in version 4.0.0 :tada:

**Comment by cbertinato** [comment]
I believe there was indeed a variable name change not mentioned in the patch notes for v4.0.0: `sqs_queue_arn` -> `queue_arn`

**Comment by bryantbiggs** [comment]
@cbertinato are you referring to whats detailed under item 5 here https://github.com/terraform-aws-modules/terraform-aws-sqs/blob/master/UPGRADE-4.0.md#variable-and-output-changes

>  All output names have had the sqs_ prefix removed

**Comment by cbertinato** [comment]
Oops! So sorry about that. Totally missed it. Apologies.

**Pull request merged**